<a href="https://colab.research.google.com/github/rajeshsompalli/Rajesh-/blob/main/MAJOR_PROJECT_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **MULTIPLE LINEAR REGRESSION USING IPL Dataset**

## **Importing the IPL data set**

In [ ]:
import pandas as pd

url="C:\datasets\IPL IMB381IPL2013.csv"  #Give path of where you have saved your csv file
ipl_df = pd.read_csv(url)

ipl_df.info()
ipl_df.shape


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 26 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Sl.NO.         130 non-null    int64  
 1   PLAYER NAME    130 non-null    object 
 2   AGE            130 non-null    int64  
 3   COUNTRY        130 non-null    object 
 4   TEAM           130 non-null    object 
 5   PLAYING ROLE   130 non-null    object 
 6   T-RUNS         130 non-null    int64  
 7   T-WKTS         130 non-null    int64  
 8   ODI-RUNS-S     130 non-null    int64  
 9   ODI-SR-B       130 non-null    float64
 10  ODI-WKTS       130 non-null    int64  
 11  ODI-SR-BL      130 non-null    float64
 12  CAPTAINCY EXP  130 non-null    int64  
 13  RUNS-S         130 non-null    int64  
 14  HS             130 non-null    int64  
 15  AVE            130 non-null    float64
 16  SR-B           130 non-null    float64
 17  SIXERS         130 non-null    int64  
 18  RUNS-C    

(130, 26)

In [ ]:
ipl_df.head()

Sl.NO.   PLAYER NAME  AGE COUNTRY  TEAM PLAYING ROLE  T-RUNS  T-WKTS  \
0       1   Abdulla, YA    2      SA  KXIP   Allrounder       0       0   
1       2  Abdur Razzak    2     BAN   RCB       Bowler     214      18   
2       3   Agarkar, AB    2     IND   KKR       Bowler     571      58   
3       4     Ashwin, R    1     IND   CSK       Bowler     284      31   
4       5  Badrinath, S    2     IND   CSK      Batsman      63       0   

   ODI-RUNS-S  ODI-SR-B  ...    SR-B  SIXERS  RUNS-C  WKTS  AVE-BL   ECON  \
0           0      0.00  ...    0.00       0     307    15   20.47   8.90   
1         657     71.41  ...    0.00       0      29     0    0.00  14.50   
2        1269     80.62  ...  121.01       5    1059    29   36.52   8.81   
3         241     84.56  ...   76.32       0    1125    49   22.96   6.23   
4          79     45.93  ...  120.71      28       0     0    0.00   0.00   

   SR-BL  AUCTION YEAR  BASE PRICE  SOLD PRICE  
0  13.93          2009       50000       50000  
1   0.00          2008       50000       50000  
2  24.90          2008      200000      350000  
3  22.14          2011      100000      850000  
4   0.00          2011      100000      800000  

[5 rows x 26 columns]

## **Displaying the initial 10 columns for the first 5 records**

In [ ]:
ipl_df.iloc[0:5, 0:10]

Sl.NO.   PLAYER NAME  AGE COUNTRY  ... T-RUNS T-WKTS  ODI-RUNS-S  ODI-SR-B
0       1   Abdulla, YA    2      SA  ...      0      0           0      0.00
1       2  Abdur Razzak    2     BAN  ...    214     18         657     71.41
2       3   Agarkar, AB    2     IND  ...    571     58        1269     80.62
3       4     Ashwin, R    1     IND  ...    284     31         241     84.56
4       5  Badrinath, S    2     IND  ...     63      0          79     45.93

[5 rows x 10 columns]

## **Predicting the players sold price**

In [ ]:
X_features = ['AGE', 'COUNTRY', 'PLAYING ROLE', 'T-RUNS', 'T-WKTS', 'ODI-RUNS-S', 'ODI-SR-B', 'ODI-WKTS', 'ODI-SR-BL','CAPTAINCY EXP', 'RUNS-S', 'HS', 'AVE', 'SR-B','SIXERS', 'RUNS-C', 'WKTS', 'AVE-BL', 'ECON', 'SR-BL']

## **Some of these features are categorical, so encoding them using one-hot encoding. A categorical feature with k values is encoded using (k-1) binary/dummy features.**

In [ ]:
categorical_features = ['AGE', 'COUNTRY','PLAYING ROLE','CAPTAINCY EXP']
ipl_encoded_df = pd.get_dummies(ipl_df[X_features], columns=categorical_features, drop_first = True)

X_features = ipl_encoded_df.columns
print(X_features)

Index(['T-RUNS', 'T-WKTS', 'ODI-RUNS-S', 'ODI-SR-B', 'ODI-WKTS', 'ODI-SR-BL',
       'RUNS-S', 'HS', 'AVE', 'SR-B', 'SIXERS', 'RUNS-C', 'WKTS', 'AVE-BL',
       'ECON', 'SR-BL', 'AGE_2', 'AGE_3', 'COUNTRY_BAN', 'COUNTRY_ENG',
       'COUNTRY_IND', 'COUNTRY_NZ', 'COUNTRY_PAK', 'COUNTRY_SA', 'COUNTRY_SL',
       'COUNTRY_WI', 'COUNTRY_ZIM', 'PLAYING ROLE_Batsman',
       'PLAYING ROLE_Bowler', 'PLAYING ROLE_W. Keeper', 'CAPTAINCY EXP_1'],
      dtype='object')


## **Adding the constant column to estimate Beta 0 using statsmodel library. Split the dataset into Train and Validation Sets**

In [ ]:
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
X = sm.add_constant(ipl_encoded_df)
Y = ipl_df['SOLD PRICE']

train_X, test_X, train_y, test_y = train_test_split(X,Y,train_size = 0.8, random_state = 42)
#random_state is set to the seed 42 randomly. This is to reproduce the result across the runs. 
#Also it ensures same records go for training and validation

C:\Users\meghu\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


## **Building the model on training set**

In [ ]:
ipl_mlr_model1 = sm.OLS(train_y,train_X).fit()
ipl_mlr_model1.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                            Results: Ordinary least squares
========================================================================================
Model:                     OLS                     Adj. R-squared:            0.362     
Dependent Variable:        SOLD PRICE              AIC:                       2965.2841 
Date:                      2021-10-21 16:35        BIC:                       3049.9046 
No. Observations:          104                     Log-Likelihood:            -1450.6   
Df Model:                  31                      F-statistic:               2.883     
Df Residuals:              72                      Prob (F-statistic):        0.000114  
R-squared:                 0.554                   Scale:                     1.1034e+11
----------------------------------------------------------------------------------------
                          Coef.       Std.Err.     t    P>|t|     [0.025       0.975]   
----------------------------------------------------------------------------------------
const                   375827.1991 228849.9306  1.6422 0.1049  -80376.7996  832031.1978
T-RUNS                     -53.7890     32.7172 -1.6441 0.1045    -119.0096      11.4316
T-WKTS                    -132.5967    609.7525 -0.2175 0.8285   -1348.1162    1082.9228
ODI-RUNS-S                  57.9600     31.5071  1.8396 0.0700      -4.8482     120.7681
ODI-SR-B                  -524.1450   1576.6368 -0.3324 0.7405   -3667.1130    2618.8231
ODI-WKTS                   815.3944    832.3883  0.9796 0.3306    -843.9413    2474.7301
ODI-SR-BL                 -773.3092   1536.3334 -0.5033 0.6163   -3835.9338    2289.3154
RUNS-S                     114.7205    173.3088  0.6619 0.5101    -230.7643     460.2054
HS                       -5516.3354   2586.3277 -2.1329 0.0363  -10672.0855    -360.5853
AVE                      21560.2760   7774.2419  2.7733 0.0071    6062.6080   37057.9439
SR-B                     -1324.7218   1373.1303 -0.9647 0.3379   -4062.0071    1412.5635
SIXERS                    4264.1001   4089.6000  1.0427 0.3006   -3888.3685   12416.5687
RUNS-C                      69.8250    297.6697  0.2346 0.8152    -523.5687     663.2187
WKTS                      3075.2422   7262.4452  0.4234 0.6732  -11402.1778   17552.6622
AVE-BL                    5182.9335  10230.1581  0.5066 0.6140  -15210.5140   25576.3810
ECON                     -6820.7781  13109.3693 -0.5203 0.6045  -32953.8282   19312.2721
SR-BL                    -7658.8094  14041.8735 -0.5454 0.5871  -35650.7726   20333.1539
AGE_2                  -230767.6463 114117.2005 -2.0222 0.0469 -458256.1279   -3279.1648
AGE_3                  -216827.0808 152246.6232 -1.4242 0.1587 -520325.1772   86671.0155
COUNTRY_BAN            -122103.5196 438719.2796 -0.2783 0.7816 -996674.4194  752467.3801
COUNTRY_ENG             672410.7654 238386.2220  2.8207 0.0062  197196.5172 1147625.0135
COUNTRY_IND             155306.4011 126316.3449  1.2295 0.2229  -96500.6302  407113.4325
COUNTRY_NZ              194218.9120 173491.9293  1.1195 0.2667 -151630.9280  540068.7521
COUNTRY_PAK              75921.7670 193463.5545  0.3924 0.6959 -309740.7804  461584.3143
COUNTRY_SA               64283.3894 144587.6773  0.4446 0.6579 -223946.8775  352513.6563
COUNTRY_SL               17360.1530 176333.7497  0.0985 0.9218 -334154.7526  368875.0586
COUNTRY_WI               10607.7792 230686.7892  0.0460 0.9635 -449257.9303  470473.4887
COUNTRY_ZIM            -145494.4793 401505.2815 -0.3624 0.7181 -945880.6296  654891.6710
PLAYING ROLE_Batsman     75724.7643 150250.0240  0.5040 0.6158 -223793.1844  375242.7130
PLAYING ROLE_Bowler      15395.8752 126308.1272  0.1219 0.9033 -236394.7744  267186.5249
PLAYING ROLE_W. Keeper  -71358.6280 213585.7444 -0.3341 0.7393 -497134.0278  354416.7718
CAPTAINCY EXP_1         164113.3972 123430.6353  1.3296 0.1878  -81941.0772  410167.8716
----------------------------------------------------------------

## **Validating this model on the test data and finding the regression parameters, r2_score and RMSE of this model1**

In [ ]:
from sklearn.metrics import r2_score, mean_squared_error
import numpy as np

pred_y = ipl_mlr_model1.predict(test_X)
print("The r2_score of the model is", np.abs(r2_score(test_y, pred_y)))
print("The RMSE of the trained model is", np.sqrt(mean_squared_error(test_y, pred_y)))
print("The model parameters are")
print(ipl_mlr_model1.params)

The r2_score of the model is 0.3191598450075729
The RMSE of the trained model is 304522.9980115466
The model parameters are
const                     375827.199059
T-RUNS                       -53.789030
T-WKTS                      -132.596723
ODI-RUNS-S                    57.959956
ODI-SR-B                    -524.144970
ODI-WKTS                     815.394424
ODI-SR-BL                   -773.309199
RUNS-S                       114.720547
HS                         -5516.335398
AVE                        21560.275955
SR-B                       -1324.721784
SIXERS                      4264.100109
RUNS-C                        69.825020
WKTS                        3075.242189
AVE-BL                      5182.933487
ECON                       -6820.778058
SR-BL                      -7658.809364
AGE_2                    -230767.646329
AGE_3                    -216827.080829
COUNTRY_BAN              -122103.519643
COUNTRY_ENG               672410.765359
COUNTRY_IND               155306.401

## **Finding the variables which are statistically significant in influencing the response variable. If the p-value of a variable is less than the significance value(alpha = 0.05), it is an influencing variables**

In [ ]:
s = ipl_mlr_model1.pvalues
# print(s.index)
impt_vars = list(s.index)  # s.index returns the variable names of the pandas series s. 
print(impt_vars)
pvals = list(s)
print(pvals)
new = [i for i in zip(impt_vars,pvals)] 
print("List of all variables and their pvalues is",new)
important_vars = [ i[0]  for i in new if i[1] < 0.05]

print("List of influencing variables is",important_vars)

['const', 'T-RUNS', 'T-WKTS', 'ODI-RUNS-S', 'ODI-SR-B', 'ODI-WKTS', 'ODI-SR-BL', 'RUNS-S', 'HS', 'AVE', 'SR-B', 'SIXERS', 'RUNS-C', 'WKTS', 'AVE-BL', 'ECON', 'SR-BL', 'AGE_2', 'AGE_3', 'COUNTRY_BAN', 'COUNTRY_ENG', 'COUNTRY_IND', 'COUNTRY_NZ', 'COUNTRY_PAK', 'COUNTRY_SA', 'COUNTRY_SL', 'COUNTRY_WI', 'COUNTRY_ZIM', 'PLAYING ROLE_Batsman', 'PLAYING ROLE_Bowler', 'PLAYING ROLE_W. Keeper', 'CAPTAINCY EXP_1']
[0.10490056300775813, 0.10452389864625473, 0.8284650387483237, 0.06995144805557298, 0.7405192506862497, 0.330571504927199, 0.6162566266912521, 0.5101209414895249, 0.036344110068239587, 0.007061202278855836, 0.3379013637272189, 0.3005898789801117, 0.8152066325801879, 0.6732332708123925, 0.613960524198236, 0.6044521679304626, 0.5871460364373555, 0.04687188990190799, 0.1587131992808342, 0.7815659102069415, 0.006187800746682858, 0.22288580255403864, 0.2666606170703112, 0.6958965700366699, 0.6579439107634066, 0.9218481195870434, 0.9634507275299442, 0.7181351775439581, 0.615805956001891, 0.9

**9. Train a new model with these significant variables**

In [ ]:
train_X = train_X[important_vars]
ipl_mlr_model2 = sm.OLS(train_y, train_X).fit()
ipl_mlr_model2.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                        Results: Ordinary least squares
================================================================================
Model:                  OLS              Adj. R-squared (uncentered): 0.655     
Dependent Variable:     SOLD PRICE       AIC:                         2977.6511 
Date:                   2021-10-21 16:36 BIC:                         2988.2287 
No. Observations:       104              Log-Likelihood:              -1484.8   
Df Model:               4                F-statistic:                 50.37     
Df Residuals:           100              Prob (F-statistic):          3.75e-23  
R-squared (uncentered): 0.668            Scale:                       1.5330e+11
---------------------------------------------------------------------------------
                Coef.       Std.Err.      t     P>|t|      [0.025        0.975]  
---------------------------------------------------------------------------------
HS            -2161.2002    2138.0051  -1.0108  0.3145    -6402.9413    2080.5410
AVE           29837.6936    6512.8158   4.5814  0.0000    16916.4525   42758.9348
AGE_2         19506.0521   67038.5770   0.2910  0.7717  -113496.5754  152508.6795
COUNTRY_ENG  347611.0068  236406.2403   1.4704  0.1446  -121412.2407  816634.2542
--------------------------------------------------------------------------------
Omnibus:                   2.563             Durbin-Watson:                2.083
Prob(Omnibus):             0.278             Jarque-Bera (JB):             2.062
Skew:                      0.332             Prob(JB):                     0.357
Kurtosis:                  3.186             Condition No.:                389  
================================================================================

"""

**Following are the inferences from the model generated:**
1. All the variables are statistically significant in predicting the outcome.
2. The R-Squared has significantly increased from the previous model. From 0.554 to 0.668.
3. The model can explain 66.8% of the variation in SOLD Price, since R-Squared is 0.668.

**10. Make Preditions on validation set of this model2**

In [ ]:
new_test_X = test_X[train_X.columns]
pred_y = ipl_mlr_model2.predict(new_test_X)

**11. Find r2_score and RMSE of this prediction by model2**

In [ ]:
from sklearn.metrics import r2_score, mean_squared_error
import numpy as np
print("The r2_score of the model is", np.abs(r2_score(test_y, pred_y)))
print("The RMSE of the trained model is", np.sqrt(mean_squared_error(test_y, pred_y)))
print("The model parameters are")
print(ipl_mlr_model2.params)

The r2_score of the model is 0.21816020642155953
The RMSE of the trained model is 407333.1164547601
The model parameters are
HS              -2161.200195
AVE             29837.693625
AGE_2           19506.052089
COUNTRY_ENG    347611.006756
dtype: float64


## References:-

https://en.wikipedia.org/wiki/Python_(programming_language)

https://towardsdatascience.com/train-validation-and-test-sets-72cb40cba9e7

https://en.wikipedia.org/wiki/Root-mean-square_deviation